CORDA-3603 Save completed flow information

williamvigorr3 opened the pull request:
When a flow is finished do not delete the checkpoint from the checkpoint table instead store the result in the checkpoint table.

williamvigorr3 committed:
Save completed flow information in the checkpoint table

Result is now stored in the database and the checkpoint is no
longer removed from the database.

williamvigorr3 committed:
Update test to also check the FlowStatus is correctly set.

williamvigorr3 committed:
Remove unnecessary non-null assertion

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/transitions/TopLevelTransition.kt:
```diff
@@ -197,17 +197,15 @@ class TopLevelTransition(
-                                        numberOfSuspends = checkpoint.checkpointState.numberOfSuspends + 1
-                            )),
+                                        numberOfSuspends = checkpoint.checkpointState.numberOfSuspends + 1),
+                                        result = event.returnValue, status = Checkpoint.FlowStatus.COMPLETED),
```
Can we format these onto new lines 😄

lankydan commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt:
```diff
@@ -285,6 +299,26 @@ class FlowFrameworkTests {
+            val checkpoint = dbCheckpointStorage.getCheckpoint(flow.id)
+            assertEquals(Checkpoint.FlowStatus.COMPLETED, checkpoint!!.status)
+            val result = checkpoint!!.result?.bytes?.deserialize<Any>()
+            assertTrue(result is Unit)
```
Can we have a few tests that store different results?

I think the `Unit` test is good (wouldn't have thought of doing that originally).

Testing what happens with the following results would be good:

- `null` from a nullable result
- `SignedTransaction`
- Any general custom object

lankydan replied:
Actually maybe, we should be doing this in `DBCheckpointStorageTest` since that is where the actual persisting happens... Can you do it there instead?

lankydan commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt:
```diff
@@ -285,6 +299,26 @@ class FlowFrameworkTests {
     }
 
+    @Test(timeout = 300_000)
+    fun `Result is stored in database when the flow finishes`() {
```
I'm considering if we should make some new test classes for these tests and tests coming in future changes.

Possibly `FlowFinishTest`. We can then have `FlowErrorTest`, `FlowStartTest`, maybe even `FlowKillTest`. I will ask for other's opinions before we do this. Just make changes to this file for now before we make a decision.

williamvigorr3 replied:
I agree that would be a good idea.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/transitions/TopLevelTransition.kt:
```diff
@@ -197,17 +197,15 @@ class TopLevelTransition(
                                 checkpointState = checkpoint.checkpointState.copy(
-                                        numberOfSuspends = checkpoint.checkpointState.numberOfSuspends + 1
-                            )),
+                                        numberOfSuspends = checkpoint.checkpointState.numberOfSuspends + 1),
```
I'm curious (I'll need to change the code), since we are no longer deleting the checkpoint, what data is held within the checkpoint's `flowState`? I'll try check this, not that it should matter since the flow is completely finished.

kyriathar replied:
Hmmmm, maybe we could remove `flowState` upon executing `Action.RemoveFlow`.

lankydan replied:
There is another task to do that separately from this PR

williamvigorr3 committed:
Fix formatting

lankydan commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt:
```diff
@@ -285,6 +299,26 @@ class FlowFrameworkTests {
+        aliceNode.database.transaction {
+            val checkpoint = dbCheckpointStorage.getCheckpoint(flow.id)
+            assertEquals(Checkpoint.FlowStatus.COMPLETED, checkpoint!!.status)
+            val result = checkpoint.result?.bytes?.deserialize<Any>()
```
Can we also go and directly access the `DBFlowResult` entity and assert that its values are what we would expect.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/transitions/TopLevelTransition.kt:
```diff
@@ -198,16 +198,17 @@ class TopLevelTransition(
                                         numberOfSuspends = checkpoint.checkpointState.numberOfSuspends + 1
-                            )),
+                                ),
+                                result = event.returnValue,
```
We can set the `result` on the checkpoint as we have done here. But we should remove/comment out the saving of the result data in `DBCheckpoint` storage.

I have got a PR approved to manually handle saving the result record. If we rebase this branch once that is merged and then you remove the code that persists the result record then that will work.

So either, don't set the `result` on the `Checkpoint` or remove the result record persisting in `DBCheckpointStorage`. Any preference @rick-r3 ?

kyriathar commented on node/src/main/kotlin/net/corda/node/services/statemachine/transitions/TopLevelTransition.kt:
```diff
@@ -197,17 +197,15 @@ class TopLevelTransition(
-                        actions.add(Action.RemoveCheckpoint(context.id))
-                    }
                     actions.addAll(arrayOf(
+                            Action.PersistCheckpoint(context.id, currentState.checkpoint, currentState.isAnyCheckpointPersisted),
```
I think: `Action.PersistCheckpoint` here seems a bit misleading, because in fact we will remove at least most basic parts of the actual checkpoint (like the fiber's stack). Perhaps we could instead keep the `RemoveCheckpoint` action and change its behaviour instead ? Moreover, If we keep it this way, means that we would also have to go and do some kind of distinction within the DBCheckpointStorage between the different `PersistCheckpoint`s now.

lankydan replied:
I think persist checkpoint is still correct. We can review this when we get onto the task that clears out the flow stack, but that is still the same in my opinion. We update the checkpoint to have no flow stack and call persist.

williamvigorr3 committed:
No longer persist the result.

I have still left in lots of tests of this. We might want to back
these changes out more later.

lankydan reviewed (changes requested):
After talking to @rick-r3 we will do the following:

- Set the result on the `Checkpoint` object
- In `DBCheckpointStorage.updateDBCheckpoint` set the entities result field to `null` in all cases (do not store the result).
- Comment out the line that creates the result entity and leave a comment that this will be added back in the future (do not use TODO or detekt will break 😄)

After this it should still work as you have done it now (only the status is updated).

We will also need to `@Ignore` the tests in `DBCheckpointStorageTest` around storing results.

williamvigorr3 committed:
Change DBCheckpointStorage to not persist the result

Statemachine still sets this on object

lankydan reviewed (dismissed):
Looks good. Two in one day?

williamvigorr3 committed:
Select on checkpoints which have not completed in many tests

williamvigorr3 committed:
Improve naming of methods I updated

williamvigorr3 committed:
Update missed test to only check for incomplete checkpoints

rick-r3 reviewed (approved):
LGTM

rick-r3 commented:
Feel free to merge @lankydan

williamvigorr3 commented:
🍾
